plugins/discovery/discovery.go

```diff
@@ -251,8 +251,39 @@ type pluginreconfig struct {
 	Plugin plugins.Plugin
 }
 
-func getPluginSet(initfuncs map[string]plugins.PluginInitFunc, manager *plugins.Manager, config *config.Config) (*pluginSet, error) {
+type pluginfactory struct {
+	name    string
+	factory plugins.Factory
+	config  interface{}
+}
+
+func getPluginSet(factories map[string]plugins.Factory, manager *plugins.Manager, config *config.Config) (*pluginSet, error) {
+
+	// Parse and validate plugin configurations.
+	pluginNames := []string{}
+	pluginFactories := []pluginfactory{}
+
+	for k := range config.Plugins {
+		f, ok := factories[k]
+		if !ok {
+			return nil, fmt.Errorf("plugin %q not registered", k)
+		}
+
+		c, err := f.Validate(manager, config.Plugins[k])
+		if err != nil {
+			return nil, err
+		}
 
+		pluginFactories = append(pluginFactories, pluginfactory{
+			name:    k,
+			factory: f,
+			config:  c,
+		})
+
+		pluginNames = append(pluginNames, k)
+	}
```

Comment:
The problem with partially activating the discovery bundle is that we don't have a good way to report status back to the control plane that there was an error. With this approach, OPA can validate all configuration first and then activate. If there's an error, the status will be reported. We already do this for bundle/status/log components but because the plugin interface didn't invalidate a validate function, we couldn't do this for the custom ones.